stingray/lightcurve.py

```diff
@@ -151,74 +183,146 @@ def __init__(self, time, counts, err=None, input_counts=True,
+            dt = np.median(np.diff(self.time))
 
+        self.dt = dt
         self.err_dist = err_dist
 
-        if unsorted:
-            self.tstart = np.min(self.time) - 0.5 * self.dt
-            self.tseg = np.max(self.time) - np.min(self.time) + self.dt
-        else:
-            self.tstart = self.time[0] - 0.5 * self.dt
-            self.tseg = self.time[-1] - self.time[0] + self.dt
-
-        self.gti = \
-            np.asarray(assign_value_if_none(gti,
-                                            [[self.tstart,
-                                              self.tstart + self.tseg]]))
-
-        check_gtis(self.gti)
-
-        good = create_gti_mask(self.time, self.gti, dt=self.dt)
+        self.tstart = self.time[0] - 0.5 * self.dt
+        self.tseg = self.time[-1] - self.time[0] + self.dt
 
-        self.time = self.time[good]
+        self.gti = gti
+        if gti is None:
+            self.gti = \
+                np.asarray([[self.tstart, self.tstart + self.tseg]])
+
+        self._mask = None
+        self._counts = None
+        self._counts_err = None
+        self._countrate = None
+        self._countrate_err = None
+        self._meanrate = None
+        self._meancounts = None
+        self._bin_lo = None
+        self._bin_hi = None
+        self._n = None
 
         if input_counts:
-            self.counts = np.asarray(counts)[good]
-            self.countrate = self.counts / self.dt
-            self.counts_err = np.asarray(err)[good]
-            self.countrate_err = np.asarray(err)[good] / self.dt
+            self._counts = np.asarray(counts)
+            self._counts_err = err
         else:
-            self.countrate = np.asarray(counts)[good]
-            self.counts = self.countrate * self.dt
-            self.counts_err = np.asarray(err)[good] * self.dt
-            self.countrate_err = np.asarray(err)[good]
-
-        self.meanrate = np.mean(self.countrate)
-        self.meancounts = np.mean(self.counts)
-        self.n = self.counts.shape[0]
-
+            self._countrate = np.asarray(counts)
+            self._countrate_err = err
+
+        if not skip_checks:
+            self.check_lightcurve()
+
+    @property
+    def mask(self):
+        if self._mask is None:
+            self._mask = create_gti_mask(self.time, self.gti, dt=self.dt)
+        return self._mask
+
+    @property
+    def n(self):
+        if self._n is None:
+            self._n = self.counts.shape[0]
+        return self._n
+
+    @property
+    def meanrate(self):
+        if self._meanrate is None:
+            self._meanrate = np.mean(self.countrate[self.mask])
+        return self._meanrate
+
+    @property
+    def meancounts(self):
+        if self._meancounts is None:
+            self._meancounts = np.mean(self.counts[self.mask])
+        return self._meancounts
+
+    @property
+    def counts(self):
+        if self._counts is None:
+            self._counts = self.countrate  * self.dt
+        return self._counts
```

Comment:
I think I need some help understanding this lazy evaluation behaviour.

So what happens if `input_counts` == True. The code above will set `_counts` with the counts array. Whenever `counts` is called, will it run the function above? Will you then have both `_counts` and `counts` set to the array? 

What happens when you then call `countrate`? Will it set `countrate == [array with count rates]` the first time it's called? Or will the property `countrate` be evaluated every time you call `countrate`?

Reply:
Every time you call `counts`, it will return `_counts` if not None, or calculate it otherwise.
Every timeyou call `countrate`, it will return `_countrate` if not None, or calculate it otherwise.
At lightcurve creation, you are forced to specify one of them, so it should be safe.

Reply:
so if I specify `counts`, and then call `countrate` a million times, will it calculate `countrate` from `counts` a million times, or will it calculate it onces and store the result? The former has the advantage of being memory-efficient, though potentially slower, the second will require more memory, but be faster?

Reply:
The second. Then, every time I _set_ counts, the following properties will be set to None again: `countrate`, `meancounts`, `meanrate` (PR to come).

Reply:
Thinking about the future with things like NICER, is that the behaviour we want? Or would it be worth taking the computational cost of just storing one (either `counts` or `countrate`) and computing the other every time it's called?

Reply:
New option: `low_memory`